### PR TITLE
[META] Add a CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by vberger**
This PR introduces a `CONTRIBUTING.md` file based on the discussion in #669. This is for now very WIP and intended to serve as a basis for discussion, like an RFC.

Regarding the most fundamental still-open questions:

- who exactly is a "maintainer" of winit, who has commit right and is willing to dedicate time to winit?
- For now I've left #252 as the "official list of features", but we might want to make it more formal maybe?

Rendered version of the file: https://github.com/vberger/winit/blob/contributing/CONTRIBUTING.md

### CODE REVIEW

**Comment by Osspial** [comment]
I feel like an official list of supported/targeted features is important enough to get its own file in root (perhaps `SCOPE.md`?).

Also, as far as tier-1 features go, there are a few things I'm not seeing on the list in the issue:
* Gyroscope support.
* Gamepad/joystick support.
* Creating child windows relative to the client area of other windows, for popup windows. This is essential for desktop applications, and is finicky without direct API access. IIRC it's even impossible to do on Wayland *without* direct API access.
* Disabling windows in favor of child windows (e.g. for save dialogs).

Where do we stand on those? AFAIK there was hesitation about gamepad support, the windowing features haven't received much attention, and gyroscope support hasn't received *any* attention despite being common on mobile devices.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# Winit Contributing Guidelines
+
+## Scope
+
+Winit aims to provide a generic platform abstracting the main graphic platforms (Windows, MacOS, x11, 
+Wayland, Android, iOS and the web platform via Emscripten).
+
+Most platform expose capabilities that cannot be meaningfully transposed to the others. Winit does not
+aim at supporting every single functionnaly of every platform, but rather abstract the set of
```
Comment (mitchmindtree) [inline CONTRIBUTING.md]: functionnaly -> functionality

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+- you tested your modifications on all the platforms impacted, or if not possible detail which platforms
+  were not tested, and what should be tested, so that a maintainer or an other contributor can test them
+- you updated any relevant documentation in winit
+- you left comments in your code explaining any part that is not straightforward, so that the 
+  maintainers and future contributors don't have to try to guess what your code is supposed to do
+
+Once your PR is open, you can ask for review by a maintainer of your platform. Winit's merging policy
+is that a PR must be approved by at least two maintainers of winit before being merged, including
+at least a maintainer of the platoform (a maintainer making a PR themselves counts as approving it).
```
Comment (mitchmindtree) [inline CONTRIBUTING.md]: platoform -> platform

**Comment by mitchmindtree** [comment]
@Osspial I like your idea of having a SCOPE.md file! It always takes me longer than I'd like to hunt down the the feature table issue. Having a dedicated file would make it easier to find, review and make updates. Perhaps we could even require that changes to it be mentioned in the CHANGELOG.md.

Re tier 1 features: Maybe we can leave this discussion for a future PR/issue dedicated to this. It might be easy to unintentionally derail @vberger's CONTRIBUTING.md if we start the discussion here :)

**Comment by vberger** [comment]
@Osspial :+1: for making a `SCOPE.md` file, in which case I suggest we could open a new PR for adding it and moving there discussion of defining the scope of winit. I feel having several discussions on the same thread would make things too complicated to follow for everybody.

@mitchmindtree Do you think it may be worth adding a third level, between `M` and `T` wrt to platform access/knowledge? So something like:

- `M` is a main maintainer for this platform
- `R` can review this platform
- `T` can test this platform

In general I think it'd be best to keep the number of levels as low as possible to keep things crystal clear, but I guess 3 levels like this is still reasonably simple.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# Winit Contributing Guidelines
+
+## Scope
+
+Winit aims to provide a generic platform abstracting the main graphic platforms (Windows, MacOS, x11, 
+Wayland, Android, iOS and the web platform via Emscripten).
+
+Most platform expose capabilities that cannot be meaningfully transposed to the others. Winit does not
```
Comment (sodiumjoe) [inline CONTRIBUTING.md]: s/platform/platforms

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+aim at supporting every single functionaly of every platform, but rather abstract the set of
```
Comment (sodiumjoe) [inline CONTRIBUTING.md]: s/aim at supporting/aim to support
s/functionaly/functionality
s/rather abstract/rather to abstract

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+Winit aims to provide a generic platform abstracting the main graphic platforms (Windows, MacOS, x11, 
+Wayland, Android, iOS and the web platform via Emscripten).
+
+Most platform expose capabilities that cannot be meaningfully transposed to the others. Winit does not
+aim at supporting every single functionaly of every platform, but rather abstract the set of 
+capabilities that is common to all platforms. In this context, APIs exposed in winit can be split into
+different "support levels":
+
+- Tier 1: features which are in the main scope of winit. They are part of the common API of winit, and
+  are taken care of by the maintainers, and them not working correctly is considered a bug in winit.
```
Comment (sodiumjoe) [inline CONTRIBUTING.md]: s/are taken care of by the maintainers, and them not working correctly is considered a bug in winit./are taken care of by the maintainers. Any part of these features that is not working correctly is considered a bug in winit.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+- Tier 2: some platform specific features can be sufficiently fundamental to the platform that winit can
```
Comment (sodiumjoe) [inline CONTRIBUTING.md]: s/platform specific/platform-specific

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  reproduce the issue, ideally by providing a minimal program exhibiting the problem
+- if it is a feature request, please provide a clear argumentation about why you believe this feature
+  should be supported by winit
+
+## Making a pull request
+
+When making a code contribution to winit, before opening your pull request, please make sure that:
+
+- you tested your modifications on all the platforms impacted, or if not possible detail which platforms
+  were not tested, and what should be tested, so that a maintainer or an other contributor can test them
```
Comment (sodiumjoe) [inline CONTRIBUTING.md]: s/an other/another

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+- you updated any relevant documentation in winit
+- you left comments in your code explaining any part that is not straightforward, so that the 
+  maintainers and future contributors don't have to try to guess what your code is supposed to do
```
Comment (sodiumjoe) [inline CONTRIBUTING.md]: also, update the changelog?

**Comment by sodiumjoe** [comment]
Left a few copy notes, I think this is great!

**Comment by Osspial** [comment]
@vberger I'm fine with splitting the scope discussion off. If we're doing that, should we move most of the current `Scope` section to `SCOPE.md`, or just replace the link with one to the new document?

I also like the idea of a reviewer tier in the maintainer/tester table.

**Comment by vberger** [comment]
@Osspial 
> I'm fine with splitting the scope discussion off. If we're doing that, should we move most of the current Scope section to SCOPE.md, or just replace the link with one to the new document?

I don't have a strong opinion on this. Probably splitting this part of to a `SCOPE.md` and just keeping a short summary in `CONTRIBUTING.md` and linking to the `SCOPE.md` file for details would be the best approach?

**Comment by Osspial** [comment]
@vberger That sounds good. For now though, I'd say to leave that part in and then have the `SCOPE.md` PR move it out.

**Comment by vberger** [comment]
@Osspial Alright, sounds good to me. Let's forward this and then open a large PR/discussion about defining the exact scope of winit.

@francesca64 Given you've been in charge of winit for some time, what's your opinion on this suggested `CONTRIBUTING.md` ?

Also, @francesca64 @mitchmindtree and @Osspial , assuming we keep the `M`/`R`/`T`/` ` levels, how should I fill in the table w.r.t. to platforms for each of you?

**Comment by Osspial** [comment]
@vberger For me, it would conservatively be this:

Windows | MacOS | x11 | Wayland | Android | iOS | Emscripten
:--: | :--: | :--: | :--: | :--: | :--: | :--:
M (I assume) | | T | T | T |  | T

Also, I don't know all that much about Emscripten but I do have the compiler set up for it and it seems to be small enough that if we need a reviewer for it learning about would be a doable task!

**Comment by mitchmindtree** [comment]
Feel free to put me down as T for X11, Wayland and Windows (though I only have access to windows via a VM).

**Comment by francesca64** [comment]
@vberger this `CONTRIBUTING.md` looks great, though I plan on providing some more detailed feedback in the next day or so.

| Windows | macOS | X11 | Wayland | Android | iOS | Emscripten |
|-|-|-|-|-|-|-|
| R | M | M |   | M | M |   |

In an ideal world, I'd only be the main maintainer for X11, so it could be good to have a note saying that macOS, Android, and iOS are still up for grabs.

@Osspial I'd be quite happy if you could take over the Emscripten backend, and I don't just say that because it's the only one I never managed to run. It's definitely the one I know the least about, and it would make the `EventLoop` 2.0 migration much less stressful for me (I also plan on reviewing your PR for that this week, since I've gotten through the rest of my backlog).

I think you'd do a great job as maintainer of the Windows backend, and I'll email @tomaka about making you a collaborator if he doesn't show up here.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+## Making a pull request
+
+When making a code contribution to winit, before opening your pull request, please make sure that:
+
+- you tested your modifications on all the platforms impacted, or if not possible detail which platforms
+  were not tested, and what should be tested, so that a maintainer or another contributor can test them
+- you updated any relevant documentation in winit
+- you left comments in your code explaining any part that is not straightforward, so that the 
+  maintainers and future contributors don't have to try to guess what your code is supposed to do
+- your PR adds an entry to the changelog file
```
Comment (francesca64) [inline CONTRIBUTING.md]: Thus far I've been enforcing this with the "if knowledge of the change could be valuable to users" proviso; do you think it's better to have everything require an entry?

Comment (rukai) [reply CONTRIBUTING.md]: IMO no. anything outside of api changes and bug fixes is just noise. Although maybe making it less subject could help. I totally stole this template line for vulkano btw, thanks.

Comment (vberger) [reply CONTRIBUTING.md]: > do you think it's better to have everything require an entry?

No, I agree with you. I just didn't think a lot about it when I wrote this.

**Comment by mtak-** [comment]
I would like to help with iOS/Android. My company has 2 devs (mostly me) actively developing a 2D game/game engine in rust using gfx-ll. I'm currently focused on iOS, and have made some contributions there (not recently). I'm intending to get rid of the longjmp hacks and port it to EventLoop 2.0 as my next task.

I'm up for maintaining iOS, and atleast able to test macos/android, but I haven't messed with that code much.

**Comment by vberger** [comment]
Given all this seems now pretty consensual, I'm going to merge this as it is now.

@Osspial Can you open the `SCOPE.md` PR, given you seem to already have some ideas for it?

@mtak- Can you then open a new PR adding yourself to the table, so that we can discuss there? I have nothing to oppose you taking the maintainership of the iOS backend, but I'd rather not delay this PR any longer. Thanks for proposing your help!

**Comment by Osspial** [comment]
@vberger Sure! It's open in #695.

**Pull request merged**